[css-anchor-position-1] Implement parsing `position-visibility`

tuankiet65 opened the pull request:
#### 86ce2d141b16d02c7c0856a8c123a50248cef321
<pre>
[css-anchor-position-1] Implement parsing `position-visibility`
<a href="https://rdar.apple.com/141799808">rdar://141799808</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=285005">https://bugs.webkit.org/show_bug.cgi?id=285005</a>

Reviewed by Antti Koivisto.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-visibility-computed-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-visibility-computed.html: Renamed from LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-visibility-computed.tentative.html.
    - Rewrite tests to match current definition in the spec, and mark it as official test.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-visibility-computed.tentative-expected.txt: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-visibility-parsing-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-visibility-parsing.html: Renamed from LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-visibility-parsing.tentative.html.
    - Rewrite tests to match current definition in the spec and mark it as official test.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-visibility-parsing.tentative-expected.txt: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/property-list.js:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-002-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-002-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
    - Define fromCSSValueID() and toCSSValueID() for PositionVisibility

* Source/WebCore/css/CSSProperties.json:
    - Add entry for position-visibility

* Source/WebCore/css/CSSValueKeywords.in:
    - Add keywords used in position-visibility

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForPositionVisibility):
    - Add function to serialize PositionVisibility
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
    - Support serializing position-visibility

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Anchor.cpp:
(WebCore::CSSPropertyParserHelpers::consumePositionVisibility):
    - Add function to parse position-visibility value

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Anchor.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::conservativelyCollectChangedAnimatableProperties const):
* Source/WebCore/rendering/style/RenderStyle.h:
    - Add initial/getter/setter functions for position-visibility

* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator&lt;&lt;):
    - Implement operator&lt;&lt;(TextStream, ) for PositionVisibility

* Source/WebCore/rendering/style/RenderStyleConstants.h:
    - Add PositionVisibility enum type

* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialPositionVisibility):
(WebCore::RenderStyle::positionVisibility const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setPositionVisibility):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
(WebCore::StyleRareNonInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
    - Add storage in RenderStyle for position-visibility
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertPositionVisibility):
    - Add function to convert CSSValue to OptionSet&lt;PositionVisibility&gt;

Canonical link: <a href="https://commits.webkit.org/291457@main">https://commits.webkit.org/291457@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f34a14bd55b134f59b0e7e01da70085ffd658a0a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/92965 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/12516 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2179 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/97964 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/43491 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95015 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/12797 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/20968 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/71083 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28497 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/95967 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/9640 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/84093 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/51411 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9333 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/1729 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/42804 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/79625 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/1703 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/99986 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20017 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/14668 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/80102 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/20268 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/79992 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79403 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19717 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23949 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/1223 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/13051 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20001 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/25177 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/19688 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23148 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/21429 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/style/StyleBuilderConverter.h:
```diff
@@ -2654,6 +2655,22 @@ inline std::optional<PositionArea> BuilderConverter::convertPositionArea(Builder
+
+    auto maybeList = requiredListDowncast<CSSValueList, CSSPrimitiveValue>(builderState, value);
+    ASSERT(maybeList);
+    auto list = *maybeList;
```
you should test and bail for every required*Downcast. 
```
if (!maybeList)
    return { };
```
This avoid Typed OM fuzzer crashes (where weird CSSValues may be constructed outside of the parsing code).

tuankiet65 replied:
Makes sense, I was under the impression only `CSSValues` from the parser can get to this method but apparently not.

tuankiet65 committed:
[css-anchor-position-1] Implement parsing `position-visibility`
rdar://141799808
https://bugs.webkit.org/show_bug.cgi?id=285005

Reviewed by Antti Koivisto.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-visibility-computed-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-visibility-computed.html: Renamed from LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-visibility-computed.tentative.html.
    - Rewrite tests to match current definition in the spec, and mark it as official test.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-visibility-computed.tentative-expected.txt: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-visibility-parsing-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-visibility-parsing.html: Renamed from LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-visibility-parsing.tentative.html.
    - Rewrite tests to match current definition in the spec and mark it as official test.

* LayoutTests/imported/w3c/web-platform-tests/css/css-anchor-position/parsing/position-visibility-parsing.tentative-expected.txt: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/property-list.js:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-002-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-002-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
    - Define fromCSSValueID() and toCSSValueID() for PositionVisibility

* Source/WebCore/css/CSSProperties.json:
    - Add entry for position-visibility

* Source/WebCore/css/CSSValueKeywords.in:
    - Add keywords used in position-visibility

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForPositionVisibility):
    - Add function to serialize PositionVisibility
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle const):
    - Support serializing position-visibility

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Anchor.cpp:
(WebCore::CSSPropertyParserHelpers::consumePositionVisibility):
    - Add function to parse position-visibility value

* Source/WebCore/css/parser/CSSPropertyParserConsumer+Anchor.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::conservativelyCollectChangedAnimatableProperties const):
* Source/WebCore/rendering/style/RenderStyle.h:
    - Add initial/getter/setter functions for position-visibility

* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator<<):
    - Implement operator<<(TextStream, ) for PositionVisibility

* Source/WebCore/rendering/style/RenderStyleConstants.h:
    - Add PositionVisibility enum type

* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::initialPositionVisibility):
(WebCore::RenderStyle::positionVisibility const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setPositionVisibility):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
(WebCore::StyleRareNonInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
    - Add storage in RenderStyle for position-visibility
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertPositionVisibility):
    - Add function to convert CSSValue to OptionSet<PositionVisibility>

Canonical link: https://commits.webkit.org/291457@main
